src/main.cpp

```diff
@@ -1083,6 +1083,18 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
                              REJECT_INVALID, "bad-txns-txouttotal-toolarge");
     }
 
+    // Check for non-zero valueBalance when there are no Sapling inputs or outputs
+    if (tx.vShieldedSpend.empty() && tx.vShieldedOutput.empty() && tx.valueBalance != 0) {
+        return state.DoS(100, error("CheckTransaction(): tx.valueBalance has no sources or sinks"),
+                            REJECT_INVALID, "bad-txns-valuebalance-nonzero");
+    }
+
+    // Check for overflow valueBalance
+    if ((tx.valueBalance & 0x7FFFFFFF) > MAX_MONEY) {
```

Comment:
valueBalance is an int64_t so this should be `0x7fffffffffffffff`

Reply:
Too many `f`s :sweat_smile:

Reply:
std::numeric_limits<int64_t>::max()